sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2466,20 +2489,26 @@ class Dataset[T] private[sql](
 
     field.dataType match {
       case IntegerType | LongType | DoubleType | FloatType | BooleanType | ByteType =>
-        val validity = allocator.buffer(numBytesOfBitmap(numOfRows))
+        val validityVector = new BitVector("validity", allocator)
+        val validityMutator = validityVector.getMutator()
+        validityVector.allocateNew(numOfRows)
+        validityMutator.setValueCount(numOfRows)
         val buf = allocator.buffer(numOfRows * field.dataType.defaultSize)
         var nullCount = 0
-        rows.foreach { row =>
+        rows.zipWithIndex.foreach { case (row, index) =>
           if (row.isNullAt(ordinal)) {
             nullCount += 1
+            validityMutator.set(index, 0)
+            fillArrow(buf, field.dataType)
```

Comment:
Yes, here are the specs of the arrow layout:
https://github.com/apache/arrow/blob/master/format/Layout.md#example-layout-int32-array